### PR TITLE
Disable 480p downscale resolution for Chromium browsers on Windows OS

### PR DESCRIPTION
**Pull request by ltrung**
**Issue #2002:**
In Chrome 98 and Windows, videos appear grayscale for 480p when there are 5-8 videos on the call (except the active speakers) and hardware acceleration is on. This issue can be mitigated by turn off hardware acceleration. 
**Description of changes:**
Check if its Chrome 98 and Windows OS then downscale to 360p for 5-8 videos instead of 480p. 

**Testing:**

*Can these tested using a demo application? Please provide reproducible step-by-step instructions.*
Yes.
- Join a meeting with 5-8 videos in Chrome 98 Windows.
- Turn on video
- Verify that the video is at 360p and appear correctly for both local attendee and remote attendees.

Test with Chrome, Edge, and Opera on Windows.

**Checklist:**

1. Have you successfully run `npm run build:release` locally?
Yes

2. Do you add, modify, or delete public API definitions? If yes, has that been reviewed and approved?
Yes. Add a helper function in browser behavior to check whether 480p needs to be disabled.

3. Do you change the wire protocol, e.g. the request method? If yes, has that been reviewed and approved?
No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/videouplinkbandwidthpolicy/NScaleVideoUplinkBandwidthPolicy.ts [hunk 1]
```diff
@@ -196,13 +198,20 @@ export default class NScaleVideoUplinkBandwidthPolicy implements VideoUplinkBand
...
         NScaleVideoUplinkBandwidthPolicy.targetHeightArray[
           Math.min(
             this.numParticipants,
             NScaleVideoUplinkBandwidthPolicy.targetHeightArray.length - 1
           )
         ];
+      if (
+        targetHeight === 480 &&
+        this.browserBehavior &&
+        this.browserBehavior.disable480pResolutionScaleDown()
```
Comment (justindarc) [inline src/videouplinkbandwidthpolicy/NScaleVideoUplinkBandwidthPolicy.ts:211]: can this be simplified as `this.browserBehavior?.disable480pResolutionScaleDown()`?

File: src/audiovideocontroller/DefaultAudioVideoController.ts [hunk 2]
```diff
@@ -526,7 +526,8 @@ export default class DefaultAudioVideoController
         this.meetingSessionContext.videoUplinkBandwidthPolicy = new NScaleVideoUplinkBandwidthPolicy(
           this.configuration.credentials.attendeeId,
           !this.meetingSessionContext.browserBehavior.disableResolutionScaleDown(),
-          this.meetingSessionContext.logger
+          this.meetingSessionContext.logger,
+          this.meetingSessionContext.browserBehavior
```
Comment (xuesichao) [inline src/audiovideocontroller/DefaultAudioVideoController.ts:530]: nit: trailing comma

File: src/browserbehavior/DefaultBrowserBehavior.ts [hunk 3]
```diff
@@ -279,6 +279,10 @@ export default class DefaultBrowserBehavior implements BrowserBehavior, Extended
     return this.isAndroid();
   }
 
+  disable480pResolutionScaleDown(): boolean {
+    return /( Chrome\/98.)/i.test(navigator.userAgent) && this.browser.os.startsWith('Windows');
```
[the remark this answers is not in the document]
Comment (ltrung) [reply src/browserbehavior/DefaultBrowserBehavior.ts]: Good catch :)

File: src/videouplinkbandwidthpolicy/NScaleVideoUplinkBandwidthPolicy.ts (hunk 1 above)
Comment (ltrung) [reply src/videouplinkbandwidthpolicy/NScaleVideoUplinkBandwidthPolicy.ts]: Thought about it but then we say we support Node 12+ and optional chaining is only available in Node 14+. Anyway, check the code and it seems we already use it everywhere else so updated :)

File: CHANGELOG.md [hunk 4]
```diff
@@ -15,6 +15,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Add support for pre-release in Versioning.
     
 ### Fixed
+- Add a workaround to avoid 480p resolution scale down for the default video uplink policy for Chromium browsers 98
```
Comment (xuesichao) [inline CHANGELOG.md:18]: Shall we mention that we scale down the resolution to 360p instead of 480p.

**Review by justindarc — APPROVED** [review]

File: src/audiovideocontroller/DefaultAudioVideoController.ts (hunk 2 above)
Comment (ltrung) [reply src/audiovideocontroller/DefaultAudioVideoController.ts:530]: ???

File: CHANGELOG.md (hunk 4 above)
Comment (ltrung) [reply CHANGELOG.md]: Updated.

File: test/videouplinkbandwidthpolicy/NScaleVideoUplinkBandwidthPolicy.test.ts [hunk 5]
```diff
@@ -488,6 +488,53 @@ describe('NScaleVideoUplinkBandwidthPolicy', () => {
...
+        true,
+        undefined,
+        new TestBrowserBehavior()
+      );
+      policy.setIdealMaxBandwidthKbps(600);
+      policy.setTransceiverController(transceiverController);
+      transceiverController.setVideoInput(new MediaStreamTrack());
+      const numParticipants = 5;
+      const expectedParams = new DefaultVideoCaptureAndEncodeParameter(
+        320,
```
Comment (LeviSklaroff) [inline test/videouplinkbandwidthpolicy/NScaleVideoUplinkBandwidthPolicy.test.ts:509]: How were these values calculated? I'm not to familiar with NScale, but 320x192 seems like an odd resolution.

Comment (ltrung) [reply test/videouplinkbandwidthpolicy/NScaleVideoUplinkBandwidthPolicy.test.ts:509]: Oh this is just a dummy test case. We do not actually use it at all. The key factor is actual the scale down resolution.

Comment (LeviSklaroff) [reply test/videouplinkbandwidthpolicy/NScaleVideoUplinkBandwidthPolicy.test.ts:509]: Ah that makes sense

**Review by LeviSklaroff — APPROVED** [review]

File: src/audiovideocontroller/DefaultAudioVideoController.ts (hunk 2 above)
Comment (xuesichao) [reply src/audiovideocontroller/DefaultAudioVideoController.ts:530]: nvm, the initialization of classes does not seem to apply this rule

**Review by xuesichao — APPROVED** [review]

File: src/browserbehavior/DefaultBrowserBehavior.ts [hunk 6]
```diff
@@ -279,6 +279,10 @@ export default class DefaultBrowserBehavior implements BrowserBehavior, Extended
     return this.isAndroid();
   }
 
+  disable480pResolutionScaleDown(): boolean {
+    return /( Chrome\/98\.)/i.test(navigator.userAgent) && this.browser.os.startsWith('Windows');
```
Comment (simmkyu) [inline src/browserbehavior/DefaultBrowserBehavior.ts:283]: Is this bug reproducible in Chromium? If yes, have we tested this user agent in Chromium on Windows, e.g., Electron?

**Pull request merged**